File: research/biomni/knowhow/workflows/scrnaseq-seurat-core-analysis/scripts/load_example_data.R
# ============================================================================
# LOAD EXAMPLE DATA
# ============================================================================
#
# Load example single-cell RNA-seq datasets for testing and demonstrations.
# This script provides user-facing functions to load data from the SeuratData
# package for learning and testing the workflow.
#
# Functions:
#   - load_seurat_data(): Load example datasets (pbmc3k, ifnb, etc.)
#
# Usage:
#   source("scripts/load_example_data.R")
#   seurat_obj <- load_seurat_data("pbmc3k")

#' Load example data from SeuratData package
#'
#' Loads publicly available single-cell RNA-seq datasets from the SeuratData
#' package. Use this for testing the workflow with real data.
#'
#' Available datasets:
#'   - "pbmc3k": 3k PBMCs from a healthy donor (10X Genomics)
#'   - "ifnb": Immune cells stimulated with interferon-beta
#'   - Other datasets from SeuratData package
#'
#' @param dataset_name Name of dataset (e.g., "pbmc3k", "ifnb")
#' @param type Type of data to load (default: auto-detect for dataset)
#' @return Seurat object with raw counts
#' @export
#'
#' @examples
#' # Load PBMC 3k dataset for testing
#' seurat_obj <- load_seurat_data("pbmc3k")
#'
#' # Load interferon-beta dataset
#' seurat_obj <- load_seurat_data("ifnb")
load_seurat_data <- function(dataset_name, type = NULL) {

  # Set CRAN mirror
  options(repos = c(CRAN = "https://cloud.r-project.org"))

  # Check if SeuratData is installed
  if (!requireNamespace("SeuratData", quietly = TRUE)) {
    message("Installing SeuratData package (first time only)...")
    if (!requireNamespace("remotes", quietly = TRUE)) {
      install.packages("remotes")
    }
    remotes::install_github('satijalab/seurat-data')
  }

  library(SeuratData)

  message("Loading ", dataset_name, " dataset from SeuratData")

  # Install dataset if not available
  if (!dataset_name %in% InstalledData()$Dataset) {
    message("Installing ", dataset_name, " dataset (this may take a few minutes)...")
    InstallData(dataset_name)
  }

  # Auto-detect type for specific datasets
  if (is.null(type)) {
    if (dataset_name == "pbmc3k") {
      type <- "default"  # pbmc3k uses "default" not "filtered"
    } else {
      type <- "filtered"
    }
  }

  # Load dataset
  LoadData(dataset_name, type = type)

  # Get the object
  seurat_obj <- get(dataset_name)

  # Update object for Seurat v5 compatibility
  if (packageVersion("Seurat") >= "5.0.0") {
    message("Updating object for Seurat v5 compatibility...")
    # SeuratObject v5 compatibility fix
    if (!".cache" %in% slotNames(seurat_obj)) {
      seurat_obj <- UpdateSeuratObject(seurat_obj)
    }
  }

  message(sprintf("✓ Data loaded successfully: %d genes x %d cells",
                  nrow(seurat_obj), ncol(seurat_obj)))

  return(seurat_obj)
}

File: data/skills/scrnaseq-scanpy-core-analysis/scripts/load_example_data.py
"""
============================================================================
LOAD EXAMPLE DATA FOR TESTING
============================================================================

This script loads example single-cell RNA-seq data for testing workflows.

Functions:
  - load_example_data(): Load PBMC 3k example dataset

Usage:
  from load_example_data import load_example_data
  adata = load_example_data()
"""

from typing import Optional


def load_example_data(dataset: str = "pbmc3k") -> 'AnnData':
    """
    Load example single-cell RNA-seq dataset.

    This function loads the PBMC 3k dataset (2,700 PBMCs from a healthy donor)
    from the scanpy datasets collection. Perfect for testing and learning.

    Parameters
    ----------
    dataset : str, optional
        Dataset to load (default: "pbmc3k")
        Options: "pbmc3k", "pbmc68k_reduced"

    Returns
    -------
    AnnData
        Example dataset with raw counts

    Examples
    --------
    >>> adata = load_example_data()
    >>> print(f"Loaded {adata.n_obs} cells and {adata.n_vars} genes")
    """
    import scanpy as sc

    print(f"Loading {dataset} example dataset...")
    print("  Source: 10X Genomics")

    if dataset == "pbmc3k":
        print("  Description: 2,700 PBMCs from a healthy donor")
        print("  Platform: 10X Chromium v1")
        adata = sc.datasets.pbmc3k()
    elif dataset == "pbmc68k_reduced":
        print("  Description: Subsampled PBMC 68k dataset")
        print("  Platform: 10X Chromium")
        adata = sc.datasets.pbmc68k_reduced()
    else:
        raise ValueError(f"Unknown dataset: {dataset}. Options: 'pbmc3k', 'pbmc68k_reduced'")

    print(f"\n✓ Data loaded successfully!")
    print(f"  Cells: {adata.n_obs}")
    print(f"  Genes: {adata.n_vars}")

    return adata


if __name__ == "__main__":
    # Test the function
    adata = load_example_data("pbmc3k")
    print("\nExample data ready for analysis!")
